### PR TITLE
Quadlet - add full support for Symlinks

### PR DESCRIPTION
**Pull request by ygalblum**
#### Does this PR introduce a user-facing change?
Yes

```release-note
Quadlet now support symbolic links also inside the search path
```

Resolves: #23755

### CODE REVIEW

File: cmd/quadlet/main.go [hunk 1]
```diff
@@ -107,31 +107,33 @@ func Debugf(format string, a ...interface{}) {
...
-	userLevelFilter := getUserLevelFilter(resolvedUnitDirAdminUser)
-
-	if rootless {
-		systemUserDirLevel := len(strings.Split(resolvedUnitDirAdminUser, string(os.PathSeparator)))
-		nonNumericFilter := getNonNumericFilter(resolvedUnitDirAdminUser, systemUserDirLevel)
-		return getRootlessDirs(nonNumericFilter, userLevelFilter)
+	// Allow overriding source dir, this is mainly for the CI tests
+	varExist, dirs := getDirsFromEnv(dirs)
+	if !varExist {
+		resolvedUnitDirAdminUser := resolveUnitDirAdminUser()
```
Comment (Luap99) [inline cmd/quadlet/main.go:116]: I this is would be cleaner written as `if varExist {return dirs}` so you do not have to indent the other logic

File: cmd/quadlet/main.go [hunk 2]
```diff
@@ -107,31 +107,33 @@ func Debugf(format string, a ...interface{}) {
...
+			dirs = getRootDirs(dirs, userLevelFilter)
+		}
 	}
 
-	return getRootDirs(userLevelFilter)
+	return dirs
 }
 
-func getDirsFromEnv() (bool, []string) {
+func getDirsFromEnv(dirs map[string]struct{}) (bool, map[string]struct{}) {
```
Comment (Luap99) [inline cmd/quadlet/main.go:131]: Unlike other types, maps are behind a shred pointer and modified in place when passed around. As such it is not needed to return the same map back to the caller unless I am missing something.

File: cmd/quadlet/main.go [hunk 3]
```diff
@@ -107,31 +107,33 @@ func Debugf(format string, a ...interface{}) {
...
+	if !varExist {
+		resolvedUnitDirAdminUser := resolveUnitDirAdminUser()
+		userLevelFilter := getUserLevelFilter(resolvedUnitDirAdminUser)
+
+		if rootless {
+			systemUserDirLevel := len(strings.Split(resolvedUnitDirAdminUser, string(os.PathSeparator)))
+			nonNumericFilter := getNonNumericFilter(resolvedUnitDirAdminUser, systemUserDirLevel)
+			dirs = getRootlessDirs(dirs, nonNumericFilter, userLevelFilter)
+		} else {
+			dirs = getRootDirs(dirs, userLevelFilter)
```
Comment (baude) [inline cmd/quadlet/main.go:124]: same idea could be used here i think? early return.

**Comment by baude** [comment]
couple of changes most of which @Luap99 points out; else LGTM

File: cmd/quadlet/main.go (hunk 1 above)
Comment (ygalblum) [reply cmd/quadlet/main.go]: You're right, fixed. With the addition of not returning `dirs` the `if` statement become even simpler

File: cmd/quadlet/main.go (hunk 2 above)
Comment (ygalblum) [reply cmd/quadlet/main.go]: Yes, you are correct. Leftover from the previous implementation. Fixed

File: cmd/quadlet/main.go (hunk 3 above)
Comment (ygalblum) [reply cmd/quadlet/main.go]: You're right. But, I have to say in the previous implementation the return inside the `rootless` statement kinda confused me. So, I rather have a single return when it's not really an early return but rather a different path

**Comment by mheon** [comment]
LGTM

**Review by giuseppe — APPROVED** [review]
/lgtm

**Pull request merged**